src/transformers/models/idefics3/processing_idefics3.py

```diff
@@ -0,0 +1,366 @@
+            truncation (`Union[bool, str, TruncationStrategy]`, *optional*):
+                Truncation strategy applied to the input ids. See [`PreTrainedTokenizerFast.truncate`] for more information.
+            max_length (`int`, *optional*):
+                Maximum length of the returned list and optionally padding/truncation length. See
+                [`PreTrainedTokenizerFast.__call__`] for more information.
+            is_split_into_words (`bool`, *optional*, defaults to `False`):
+                Whether the input text is split into words or not. If set to `True`, the tokenizer will skip the
+                tokenization process and assume the input is already tokenized.
+            add_special_tokens (`bool`, *optional*, defaults to `True`):
+                Whether to add special tokens or not. See [`PreTrainedTokenizerFast.__call__`] for more information.
+            return_tensors (`Union[str, TensorType]`, *optional*):
+                If set, will return tensors of a particular framework. See [`PreTrainedTokenizerFast.__call__`] for more
+                information.
+        """
+        if text is None and images is None:
+            raise ValueError("You must provide either `text` or `images`.")
+
+        output_kwargs = self._merge_kwargs(
+            Idefics3ProcessorKwargs,
+            tokenizer_init_kwargs=self.tokenizer.init_kwargs,
+            **kwargs,
+        )
+
+        # Temporary fix for "padding_side" in init_kwargs
+        _ = output_kwargs["text_kwargs"].pop("padding_side", None)
+
+        image_seq_len = output_kwargs["images_kwargs"].pop("image_seq_len", None)
+        image_seq_len = image_seq_len if image_seq_len is not None else self.image_seq_len
+
+        n_images_in_text = []
+        n_images_in_images = []
+        inputs = BatchFeature()
+
+        if images is not None:
+            if is_image_or_image_url(images):
+                images = [[images]]
+            elif isinstance(images, list) and is_image_or_image_url(images[0]):
+                images = [images]
+            elif (
+                not isinstance(images, list)
+                and not isinstance(images[0], list)
+                and not is_image_or_image_url(images[0][0])
+            ):
+                raise ValueError(
+                    "Invalid input images. Please provide a single image or a list of images or a list of list of images."
+                )
+            n_images_in_images = [len(sample) for sample in images]
+
+            # Load images if they are URLs
+            new_images = []
+            for sample in images:
+                new_images.append([])
+                for im in sample:
+                    if is_valid_image(im):
+                        new_images[-1].append(im)  # already loaded
+                    elif isinstance(im, str):
+                        new_images[-1].append(load_image(im))
+
+            images = new_images
+            del new_images
```

Comment:
A few comments
* appending an empty list, then indexing into it an appending again is hacky. It's cleaner to create another list you append images to and then append that list into `new_images` at the end
* Passing actual images in should be the default behaviour. We kind of let idefics get away with passing in urls, but this isn't standard and actual images should be the assumed default i.e. we should test if something is a url rather than test if it's a valid image 
* I've seen in this PR a lot the pattern of creating a temporary list, assigning it and then deleting. This isn't a good pattern - we don't want to have to keep creating and destroying objects. 

```suggestion
            images = [[load_image(im) if is_url(im) else im for im in sample] for sample in images]
```